### ISSUE TITLE
mpool allocation failed after defrag twice...

### ISSUE THREAD

**Issue by nashif** [zephyriot/zep-jira14#955]
**_Reported by Sharron LIU:_**

memory pool allocation failed after defrag twice. 

Below code test memory pool with unified kernel.
When the code is invoked from main task >> allocation failed after the 2nd defrag.
When same code invoked from Fiber >> allocation success

The test code posted, complete test app see attachment.
```
#define MPOOL_BLK_SIZE_MIN 4
#define MPOOL_BLK_SIZE_MAX 16
#define MPOOL_BLK_NUM_MIN 8
#define MPOOL_BLK_NUM_MAX 2
K_MEMORY_POOL_DEFINE(kpool, MPOOL_BLK_SIZE_MIN, MPOOL_BLK_SIZE_MAX, MPOOL_BLK_NUM_MAX);

/** @brief test memory pool functionality */
int test_bat_mpool(int argc, char *argv[])
{
        int ret = TC_PASS;
        TC_PRINT("%s\n", __FUNCTION__);

        struct k_block block[MPOOL_BLK_NUM_MIN];
        /**TESTPOINT: verify alloc*/
        /*allocated 8 blocks, each 4k*/
        for (int i = 0; i < MPOOL_BLK_NUM_MIN; i++) {
                if (k_mem_pool_alloc(&kpool, &block[i], MPOOL_BLK_SIZE_MIN,
                                K_FOREVER) != RC_OK) {
                        ret = TC_FAIL;
                        TC_ERROR("mpool: min size allocation failed %d\n", i);
                }
        }

        /**TESTPOINT: verify free*/
        /*free blocks 0,1,2,3*/
        for (int i = 0; i < (MPOOL_BLK_NUM_MIN>>1); i++) {
                k_mem_pool_free(&block[i]);
        }

        /**TESTPOINT: verify implicit defrag*/
        /*request 16k, the pool has "adjacent free blocks" to merge*/
        if (k_mem_pool_alloc(&kpool, &block[0], MPOOL_BLK_SIZE_MAX,
                        K_FOREVER) != RC_OK) {
                ret = TC_FAIL;
                TC_ERROR("mpool: implicit defragmentation failed\n");
        }
        /*free blocks 4,5,6,7*/
        for (int i = (MPOOL_BLK_NUM_MIN>>1); i < MPOOL_BLK_NUM_MIN; i++) {
                k_mem_pool_free(&block[i]);
        }

        /**TESTPOINT: verify manual defrag*/
        /*do manual defragment*/
        k_mem_pool_defrag(&kpool);
        /*request 16k, the pool is defragmented*/
        if (k_mem_pool_alloc(&kpool, &block[1], MPOOL_BLK_SIZE_MAX,
                        2000) != RC_OK) {
                ret = TC_FAIL;
                TC_ERROR("mpool: explicit defragmentation failed\n");
        }
        /*free blocks 0,1*/
        if (ret != TC_FAIL) {
                for (int i = 0; i < MPOOL_BLK_NUM_MAX; i++) {
                        k_mem_pool_free(&block[i]);
                }
        }

        return ret;
}
```

how-to-reproduce
```
# source zephyr-env.sh
# tar -zxvf test_pool_v2.tgz
# make -C test_pool_v2 qemu
# verify console output "FAIL - test_bat_mpool@69. mpool: explicit defragmentation failed"
```
When using kernel v1, the similar codes passed.

(Imported from Jira ZEP-1051)

**Comment by nashif** [comment]
**_by Sharron LIU:_**

* [test_pool_v2.tgz](https://s3.amazonaws.com/logs.zephyrproject.org/jira/GH-955/test_pool_v2.tgz) (size: 1787)

**Comment by nashif** [comment]
**_by Sharron LIU:_**

"kernel/unified/mem_pool.c:get_block_recursive()" returns NULL after the 2nd defrag.

**Comment by nashif** [comment]
**_by Sharron LIU:_**

I have the test codes written in kernel v1 API. It passed with kernel V1, but also failed with kernel V2 (legacy API). So I think this could be kernel V2 specific issue.

**Comment by nashif** [comment]
**_by Allan Stephens:_**

I've identified an issue with the kernel V2 version of memory pools, and will be submitting a fix shortly. (The defragmentation algorithm was modified to improve its readability, but this also changed the behavior of the algorithm and made it operate incorrectly in some cases.)

**Comment by nashif** [comment]
**_by Allan Stephens:_**

Please note that the test code no longer builds successfully under the latest unified kernel code. The issue is that the K_THREAD_DEFINE() macro no longer generates an "extern" for a thread's entry point function. The test code will need to add an explicit "extern" for the entry point routine before calling K_THREAD_DEFINE(), as is the normal convention when writing C code (i.e. you have to declare a function before you can use it).

Just thought I'd mention this to save Sharron some time ...

**Comment by nashif** [comment]
**_by Benjamin Walsh:_**

Fixed by be31524

**Comment by nashif** [comment]
**_by Mark Linkmeyer:_**

Fixing incorrect priority
